### DISCUSSION TITLE
Errors are coming while installing Counterfit

### DISCUSSION THREAD

**Discussion by InamulRehman** [microsoft/IoT-For-Beginners#240] [Q&A]
Hi! this is Inamulrehman,
 I am a student and i want to try all of the projects in IoT Beginners course but while i was setting up a virtual hardware environment all was going properly except Counterfit. I am unable to install its packages (while on running **pip install CounterFit**) getting some kind of goods and bads.
so please be sure me this is  available or not if yes then what i do for successful installation.

**Comment by jimbobbennett** [comment]
What errors are you getting? What version of Python are you using?

**Reply by InamulRehman** [reply]
According to instructions (in **virtual-device.md**) i installed latest python version **Python 3.9.6** . But now running command python --version in cmd gives **python 3.8.9** probably because of setting up the virtual device.
By the while doing this task **configure a Python virtual environment** i follow instructions accordingly and instep of activation command  **.venv\Scripts\activate.bat** was not found. I mean there was no scripts folder in the .venv folder. I used this command **.venv\bin\Activate.ps1** for activation and also ran the command **pip install CounterFit** in bin folder as all pip packages were in this folder, after that i was getting this kind of error

![Screenshot 2021-07-20 092604](https://user-images.githubusercontent.com/80153613/126262142-409c48a6-06f7-461f-9d66-aebdf36478ec.png)
![Screenshot 2021-07-20 092650](https://user-images.githubusercontent.com/80153613/126262146-9493b591-960f-4223-a5ed-b15c9217edb9.png)
![Screenshot 2021-07-20 092754](https://user-images.githubusercontent.com/80153613/126262147-19a2eee2-a193-485a-8634-9980ea89a4e3.png)
stuff like that.

While on vscode terminal running Counterfit gives:
![Screenshot 2021-07-20 093343](https://user-images.githubusercontent.com/80153613/126262544-b5f400c5-a134-4083-b530-39877b9a4b97.png)

**Reply by jimbobbennett** [reply]
You shouldn't be running this from inside the `.venv` folder - it should be from the folder you were in when you created the venv.

It also looks like you are using PowerShell, not the command prompt. Can you try again using cmd instead of PS?

I haven't tried this with PowerShell as I mainly work on macOS. I'll try to find some time to test it out in PS and add to the guides if there are different steps.

**Reply by InamulRehman** [reply]
**same....**
![Screenshot 2021-07-21 153010](https://user-images.githubusercontent.com/80153613/126475788-84de0786-0b37-4210-bf91-06affdb5ee52.png)
![Screenshot 2021-07-21 153517](https://user-images.githubusercontent.com/80153613/126475798-001587f1-ebc1-4efa-8410-1e4d4280abe1.png)
![Screenshot 2021-07-21 153601](https://user-images.githubusercontent.com/80153613/126475803-193cf766-a275-4d63-9a53-803989fc1460.png)

**Comment by jimbobbennett** [comment]
I've added more info to the guides when using Windows with PowerShell.

The commands to run to create the nightlight folder, set up the virtual environment, activate it, then install counterfit are:

```powershell
mkdir nightlight
cd nightlight
python3 -m venv .venv
.\.venv\Scripts\Activate.ps1
pip install CounterFit
pip install counterfit-connection
pip install counterfit-shims-grove
```


This is the output I get when doing this:

```output
PS C:\Users\jabenn> mkdir nightlight


    Directory: C:\Users\jabenn


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         7/24/2021  12:11 PM                nightlight


PS C:\Users\jabenn> cd nightlight
PS C:\Users\jabenn\nightlight> python3 -m venv .venv
PS C:\Users\jabenn\nightlight> .\.venv\Scripts\Activate.ps1
(.venv) PS C:\Users\jabenn\nightlight> pip install CounterFit
Collecting CounterFit
  Using cached CounterFit-0.1.3.dev29-py3-none-any.whl (119 kB)
Collecting beautifulsoup4==4.9.3
  Using cached beautifulsoup4-4.9.3-py3-none-any.whl (115 kB)
Collecting Flask-SocketIO==5.0.1
  Using cached Flask_SocketIO-5.0.1-py2.py3-none-any.whl (16 kB)
Collecting Flask==1.1.2
  Using cached Flask-1.1.2-py2.py3-none-any.whl (94 kB)
Collecting eventlet==0.30.2
  Using cached eventlet-0.30.2-py2.py3-none-any.whl (224 kB)
Collecting lxml==4.6.3
  Using cached lxml-4.6.3-cp38-cp38-win_amd64.whl (3.5 MB)
Collecting soupsieve>1.2
  Using cached soupsieve-2.2.1-py3-none-any.whl (33 kB)
Collecting six>=1.10.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting greenlet>=0.3
  Using cached greenlet-1.1.0-cp38-cp38-win_amd64.whl (96 kB)
Collecting dnspython<2.0.0,>=1.15.0
  Using cached dnspython-1.16.0-py2.py3-none-any.whl (188 kB)
Collecting Jinja2>=2.10.1
  Using cached Jinja2-3.0.1-py3-none-any.whl (133 kB)
Collecting Werkzeug>=0.15
  Using cached Werkzeug-2.0.1-py3-none-any.whl (288 kB)
Collecting click>=5.1
  Using cached click-8.0.1-py3-none-any.whl (97 kB)
Collecting itsdangerous>=0.24
  Using cached itsdangerous-2.0.1-py3-none-any.whl (18 kB)
Collecting python-socketio>=5.0.2
  Using cached python_socketio-5.3.0-py2.py3-none-any.whl (53 kB)
Collecting colorama
  Using cached colorama-0.4.4-py2.py3-none-any.whl (16 kB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.0.1-cp38-cp38-win_amd64.whl (14 kB)
Collecting bidict>=0.21.0
  Using cached bidict-0.21.2-py2.py3-none-any.whl (37 kB)
Collecting python-engineio>=4.1.0
  Using cached python_engineio-4.2.0-py2.py3-none-any.whl (51 kB)
Installing collected packages: MarkupSafe, colorama, Werkzeug, python-engineio, Jinja2, itsdangerous, click, bidict, soupsieve, six, python-socketio, greenlet, Flask, dnspython, lxml, Flask-SocketIO, eventlet, beautifulsoup4, CounterFit
Successfully installed CounterFit-0.1.3.dev29 Flask-1.1.2 Flask-SocketIO-5.0.1 Jinja2-3.0.1 MarkupSafe-2.0.1 Werkzeug-2.0.1 beautifulsoup4-4.9.3 bidict-0.21.2 click-8.0.1 colorama-0.4.4 dnspython-1.16.0 eventlet-0.30.2 greenlet-1.1.0 itsdangerous-2.0.1 lxml-4.6.3 python-engineio-4.2.0 python-socketio-5.3.0 six-1.16.0 soupsieve-2.2.1
WARNING: You are using pip version 21.1.1; however, version 21.1.3 is available.
You should consider upgrading via the 'c:\users\jabenn\nightlight\.venv\scripts\python.exe -m pip install --upgrade pip' command.
(.venv) PS C:\Users\jabenn\nightlight> pip install counterfit-connection
Collecting counterfit-connection
  Downloading counterfit_connection-0.1.0.dev5-py3-none-any.whl (2.7 kB)
Collecting requests
  Downloading requests-2.26.0-py2.py3-none-any.whl (62 kB)
     |████████████████████████████████| 62 kB 330 kB/s
Collecting charset-normalizer~=2.0.0
  Downloading charset_normalizer-2.0.3-py3-none-any.whl (35 kB)
Collecting idna<4,>=2.5
  Downloading idna-3.2-py3-none-any.whl (59 kB)
     |████████████████████████████████| 59 kB 1.8 MB/s
Collecting certifi>=2017.4.17
  Downloading certifi-2021.5.30-py2.py3-none-any.whl (145 kB)
     |████████████████████████████████| 145 kB 6.4 MB/s
Collecting urllib3<1.27,>=1.21.1
  Downloading urllib3-1.26.6-py2.py3-none-any.whl (138 kB)
     |████████████████████████████████| 138 kB 6.8 MB/s
Installing collected packages: urllib3, idna, charset-normalizer, certifi, requests, counterfit-connection
Successfully installed certifi-2021.5.30 charset-normalizer-2.0.3 counterfit-connection-0.1.0.dev5 idna-3.2 requests-2.26.0 urllib3-1.26.6
WARNING: You are using pip version 21.1.1; however, version 21.1.3 is available.
You should consider upgrading via the 'c:\users\jabenn\nightlight\.venv\scripts\python.exe -m pip install --upgrade pip' command.
(.venv) PS C:\Users\jabenn\nightlight> pip install counterfit-shims-grove
Collecting counterfit-shims-grove
  Downloading counterfit_shims_grove-0.1.4.dev5-py3-none-any.whl (7.1 kB)
Requirement already satisfied: counterfit-connection in c:\users\jabenn\nightlight\.venv\lib\site-packages (from counterfit-shims-grove) (0.1.0.dev5)
Requirement already satisfied: requests in c:\users\jabenn\nightlight\.venv\lib\site-packages (from counterfit-shims-grove) (2.26.0)
Requirement already satisfied: charset-normalizer~=2.0.0 in c:\users\jabenn\nightlight\.venv\lib\site-packages (from requests->counterfit-shims-grove) (2.0.3)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\jabenn\nightlight\.venv\lib\site-packages (from requests->counterfit-shims-grove) (1.26.6)
Requirement already satisfied: idna<4,>=2.5 in c:\users\jabenn\nightlight\.venv\lib\site-packages (from requests->counterfit-shims-grove) (3.2)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\jabenn\nightlight\.venv\lib\site-packages (from requests->counterfit-shims-grove) (2021.5.30)
Installing collected packages: counterfit-shims-grove
Successfully installed counterfit-shims-grove-0.1.4.dev5
WARNING: You are using pip version 21.1.1; however, version 21.1.3 is available.
You should consider upgrading via the 'c:\users\jabenn\nightlight\.venv\scripts\python.exe -m pip install --upgrade pip' command.
```

I can then run CounterFit with the following command:

```powershell
CounterFit
```

Giving this output:

```output
(.venv) PS C:\Users\jabenn\nightlight> CounterFit
CounterFit - virtual IoT hardware running on port 5000
```

and launching CounterFit in my default browser.

**Reply by jimbobbennett** [reply]
Also note that you never need to navigate into the `.venv` folder, you should do everything from the root folder that you were in when you created the virtual environment.

**Comment by InamulRehman** [comment]
Ok thanks for helping me out with this.
I will try this and will notify as soon as possible!

**Comment by InamulRehman** [comment]
I followed commands but not still able to install .
There is no path like this  **.\.venv\Scripts\Activate.ps1**    (By the at start of path there is slash between two dots which is not showing here )
I have tried by navigating to **bin** folder and see errors.
What i have done is :
![Screenshot 2021-07-28 153447](https://user-images.githubusercontent.com/80153613/127308966-7c058121-3e8e-4db9-9a03-affbc7bcdd9c.png)

**Reply by jimbobbennett** [reply]
If you are using the command prompt instead of powershell then you need to run Activate.bat, not Activate.ps1
